lib/ai/model/agent.go

```diff
@@ -0,0 +1,325 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package model
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sashabaranov/go-openai"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	actionFinalAnswer = "Final Answer"
+	actionException   = "_Exception"
+	maxIterations     = 15
+	maxElapsedTime    = 5 * time.Minute
```

Comment:
From UX perspective, 5 minutes looks like too long. But I don't know if this is a decent timer already given how our agent works.

Reply:
It's long, but we plan to keep sending updates to the WebUI, so instead of "just waiting 5min", the user will get updates on what is happening in the background.

Reply:
That's better. Thanks.